michbl/assignment2/BinaryTree.hpp

```diff
@@ -0,0 +1,194 @@
+#ifndef BINARYTREE_HPP
+#define BINARYTREE_HPP
+
+#include <iostream>
+using namespace std;
+
+// Assuming T is comparable and printable
+// Assuming keys are unique
+
+template <typename T>
+struct Node {
+    T key;
+    Node<T> *left = nullptr;
+    Node<T> *right = nullptr;
+};
+
+template <typename T>
+class BinaryTree {
+private:
+    Node<T> *root;
+
+    void copyTree(Node<T> *parent, Node<T> *newNode, string side) {
+        if (newNode == nullptr) {
+            return;
+        }
+
+        Node<T> *child = new Node<T>;
+        child->key = newNode->key;
+        if (side == "left") {
+            parent->left = child;
+        }
+        else {
+            parent->right = child;
+        }
+        copyTree(child, newNode->left, "left");
+        copyTree(child, newNode->right, "right");
+    }
+
+    void insert(Node<T> *node, T key, T parentKey, bool &success) {
+        if (node == nullptr) {
+            return;
+        }
+        if (node->key == parentKey) {
+            if (node->left == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->left = child;
+                success = true;
+                return;
+            }
+            else if (node->right == nullptr) {
+                Node<T> *child = new Node<T>;
+                child->key = key;
+                node->right = child;
+                success = true;
+                return;
+            }
+            else { // parent already has 2 children
+                return;
+            }
+        }
+        insert(node->left, key, parentKey, success);
+        if (success) {
+            return;
+        }
+        insert(node->right, key, parentKey, success);
+    }
+
+    void destroyTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        destroyTree(node->left);
+        destroyTree(node->right);
+        delete node;
+    }
+
+    void printTree(Node<T> *node) {
+        if (node == nullptr) {
+            return;
+        }
+        printTree(node->left);
+        printTree(node->right);
+        cout << node->key << " ";
+    }
+
+    void printAncestors(Node<T> *node, T key, bool &found) {

```

Comment:
You do not change `node` in this function, correct?
So you can pass `node` as a const pointer.
`void printAncestors(const Node<T> *node, T key, bool &found)`
This checks in compile time that `node` is not changed; but also provides better documentation for the user of this function.